File: integration_tests/dart_client_tests_real/test/real_server_adapter.dart
import 'dart:async';
import 'dart:convert';
import 'dart:io';
import 'package:graphlink_client_integration_tests/generated/graphlink.dart';
import 'package:http/http.dart' as http;

const _defaultPort = 9999;
const _uploadPort = 9998;

String _httpUrl(int port) => 'http://localhost:$port/graphql';
String _wsUrl(int port) => 'ws://localhost:$port/graphql';

Future<String> Function(String) httpAdapterForPort(int port) =>
    (String payload) async {
      final response = await http.post(
        Uri.parse(_httpUrl(port)),
        headers: {'Content-Type': 'application/json'},
        body: payload,
      );
      return response.body;
    };

Future<String> realHttpAdapter(String payload) =>
    httpAdapterForPort(_defaultPort)(payload);

/// Multipart adapter for the upload server on port 9998.
/// Parts map: 'operations' and 'map' are Strings; numbered keys are GLUpload.
Future<String> realMultipartAdapter(
    Map<String, Object> parts, UploadProgressCallback? onProgress) async {
  final request =
      http.MultipartRequest('POST', Uri.parse(_httpUrl(_uploadPort)));

  for (final entry in parts.entries) {
    final value = entry.value;
    if (value is String) {
      request.fields[entry.key] = value;
    } else if (value is GLUpload) {
      final bytes = await value.stream.fold<List<int>>(
        [],
        (acc, chunk) => acc..addAll(chunk),
      );
      request.files.add(http.MultipartFile.fromBytes(
        entry.key,
        bytes,
        filename: value.filename ?? 'upload',
      ));
    }
  }

  final streamed = await request.send();
  return await streamed.stream.transform(utf8.decoder).join();
}

GraphLinkClient newClient({int port = _defaultPort}) => GraphLinkClient(
      adapter: httpAdapterForPort(port),
      wsAdapter: DefaultGraphLinkWebSocketAdapter(url: _wsUrl(port)),
    );

GraphLinkClient newUploadClient() => GraphLinkClient(
      adapter: realHttpAdapter,
      uploadAdapter: realMultipartAdapter,
      wsAdapter: DefaultGraphLinkWebSocketAdapter(url: _wsUrl(_defaultPort)),
    );

DefaultGraphLinkWebSocketAdapter newWsAdapter({int port = _defaultPort}) =>
    DefaultGraphLinkWebSocketAdapter(url: _wsUrl(port), reconnect: true);

/// Adapter with maxReconnectAttempts=null (retry forever) and a reduced
/// maxReconnectDelay so integration tests run fast.
DefaultGraphLinkWebSocketAdapter newWsAdapterNeverStop(
        {int port = _defaultPort}) =>
    DefaultGraphLinkWebSocketAdapter(
      url: _wsUrl(port),
      reconnect: true,
      maxReconnectAttempts: null,
      maxReconnectDelay: const Duration(seconds: 2),
    );

// ---------------------------------------------------------------------------
// Server lifecycle utilities
// ---------------------------------------------------------------------------

String _jarPath() =>
    '${Directory.current.path}/../spring_server/build/libs/graphlink-test-server-0.0.1.jar';

/// Kills any process listening on [port].
Future<void> killPort(int port) async {
  await Process.run('bash', [
    '-c',
    'lsof -ti :$port -sTCP:LISTEN | xargs kill -9 2>/dev/null || true'
  ]);
  await Future.delayed(const Duration(milliseconds: 500));
}

/// Kills any process on [port], starts the test-server JAR on that port,
/// and waits until the port accepts TCP connections.
Future<void> startServer(int port) async {
  print("Killing process with port $port");
  await killPort(port);
  print("starting spring server with port ${port}");
  final process = await Process.start('java', [
    '-jar',
    _jarPath(),
    '--server.port=$port',
    '--management.server.port=-1',
  ]);
  unawaited(process.stdout.drain());
  unawaited(process.stderr.drain());
  print("waiting for port ${port}");
  await _waitForPort(port);
}

Future<void> _waitForPort(int port) async {
  final deadline = DateTime.now().add(const Duration(seconds: 20));
  while (DateTime.now().isBefore(deadline)) {
    try {
      final socket = await Socket.connect('localhost', port,
          timeout: const Duration(seconds: 1));
      await socket.close();
      return;
    } catch (_) {
      await Future.delayed(const Duration(milliseconds: 500));
    }
  }
  throw TimeoutException('Server on port $port did not start within 20 s');
}

File: integration_tests/typescript_client_tests_real/test/real-server-adapter.ts
import { execSync, spawn } from 'node:child_process';
import { createConnection } from 'node:net';
import { GraphLinkClient } from '../lib/generated/client/graph-link-client.js';
import type { GLUpload, GLMultipartAdapter, UploadProgressCallback } from '../lib/generated/client/graph-link-uploads.js';
import { DefaultGraphLinkWsAdapter } from '../lib/generated/graphlink.js';


const HTTP_URL = 'http://localhost:9988/graphql';
const UPLOAD_URL = 'http://localhost:9998/graphql';
export const WS_URL = 'ws://localhost:9988/graphql';

export function httpUrlForPort(port: number): string {
  return `http://localhost:${port}/graphql`;
}

export function wsUrlForPort(port: number): string {
  return `ws://localhost:${port}/graphql`;
}

export function httpAdapterForPort(port: number) {
  return async (payload: string): Promise<string> => {
    const response = await fetch(httpUrlForPort(port), {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: payload,
    });
    return response.text();
  };
}

export function newWsAdapterForPort(port: number): DefaultGraphLinkWsAdapter {
  return new DefaultGraphLinkWsAdapter(wsUrlForPort(port));
}

/** Adapter with maxReconnectAttempts=null (retry forever) and maxReconnectDelay=2s. */
export function newWsAdapterNeverStop(port: number): DefaultGraphLinkWsAdapter {
  return new DefaultGraphLinkWsAdapter(wsUrlForPort(port), undefined, null, 2_000);
}

// ---------------------------------------------------------------------------
// Server lifecycle utilities
// ---------------------------------------------------------------------------

const JAR_PATH = new URL('../../spring_server/build/libs/graphlink-test-server-0.0.1.jar', import.meta.url).pathname;

export function killPort(port: number): void {
  try {
    execSync(`lsof -ti :${port} -sTCP:LISTEN | xargs kill -9 2>/dev/null || true`, { shell: '/bin/bash' });
  } catch { /* ignore */ }
}

function waitForPort(port: number, timeoutMs = 30_000): Promise<void> {
  return new Promise((resolve, reject) => {
    const deadline = Date.now() + timeoutMs;
    function attempt() {
      const sock = createConnection(port, 'localhost');
      sock.once('connect', () => { sock.destroy(); resolve(); });
      sock.once('error', () => {
        sock.destroy();
        if (Date.now() >= deadline) { reject(new Error(`Port ${port} did not open within ${timeoutMs}ms`)); return; }
        setTimeout(attempt, 500);
      });
    }
    attempt();
  });
}

export async function startServer(port: number): Promise<void> {
  killPort(port);
  const proc = spawn('java', ['-jar', JAR_PATH, `--server.port=${port}`, '--management.server.port=-1'], { stdio: 'ignore', detached: true });
  proc.unref();
  await waitForPort(port);
}

export async function realHttpAdapter(payload: string): Promise<string> {
  const response = await fetch(HTTP_URL, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: payload,
  });
  return response.text();
}

export const realMultipartAdapter: GLMultipartAdapter = async (
  parts: Record<string, unknown>,
  _onProgress?: UploadProgressCallback,
): Promise<string> => {
  const formData = new FormData();
  for (const [key, value] of Object.entries(parts)) {
    if (typeof value === 'string') {
      formData.append(key, value);
    } else {
      const upload = value as GLUpload;
      const blob =
        upload.stream instanceof Blob
          ? upload.stream
          : new Blob([await new Response(upload.stream as ReadableStream).arrayBuffer()], {
              type: upload.mimeType,
            });
      formData.append(key, blob, upload.filename ?? 'upload');
    }
  }
  const response = await fetch(UPLOAD_URL, { method: 'POST', body: formData });
  return response.text();
};

export function newWsAdapter(): DefaultGraphLinkWsAdapter {
  return new DefaultGraphLinkWsAdapter(WS_URL);
}

export function newClient(): GraphLinkClient {
  return new GraphLinkClient(realHttpAdapter, newWsAdapter());
}

export function newUploadClient(): GraphLinkClient {
  return new GraphLinkClient(realHttpAdapter, newWsAdapter(), realMultipartAdapter);
}
